ui/contexts/snaps/snap-interface.tsx

```diff
@@ -175,7 +189,71 @@ export const SnapInterfaceContextProvider: FunctionComponent<
 
     internalState.current = state;
     updateStateDebounced(state);
-    handleInputChangeDebounced(name, value ?? '');
+    handleInputChangeDebounced(name, value);
+  };
+
+  const uploadFile = (name: string, file: File | null) => {
+    handleSnapRequest({
+      snapId,
+      origin: '',
+      handler: 'onUserInput',
+      request: {
+        jsonrpc: '2.0',
+        method: ' ',
+        params: {
+          event: {
+            type: UserInputEventType.FileUploadEvent,
+            name,
+            file,
+          },
+          id: interfaceId,
+          context,
+        },
+      },
+    }).then(() => forceUpdateMetamaskState(dispatch));
+  };
+
+  /**
+   * Handle the file change of an input.
+   *
+   * @param name - The name of the input.
+   * @param file - The file to upload.
+   * @param form - The name of the form containing the input.
+   */
+  const handleFileChange: HandleFileChange = (name, file, form) => {
+    if (file) {
+      file
+        .arrayBuffer()
+        .then((arrayBuffer) => new Uint8Array(arrayBuffer))
+        .then((uint8Array) => encodeBase64(uint8Array))
+        .then((base64) => {
+          const fileObject: FileObject = {
+            name: file.name,
+            size: file.size,
+            contentType: file.type,
+            contents: base64,
+          };
+
+          const state = mergeValue(
+            internalState.current,
+            name,
+            fileObject,
+            form,
+          );
+
+          internalState.current = state;
+          updateStateDebounced(state);
+          uploadFile(name, fileObject);
```

Comment:
Let's mirror `handleEvent` then and flush the state update:
```suggestion
          internalState.current = state;
          updateStateDebounced(state);
          updateStateDebounced.flush();
          uploadFile(name, fileObject);
```